src/firmware/guest/types/snp.rs

```diff
@@ -117,14 +113,76 @@ bitfield! {
 /// migration agent associated with it, the REPORT_ID_MA is filled in with the report ID of the
 /// migration agent.
 ///
-/// The firmware signs the attestation report with its VCEK. The firmware uses the system wide
+/// The firmware signs the attestation report with its VEK (VCEK or VLEK). The firmware uses the system wide
 /// ReportedTcb value as the TCB version to derive the VCEK. This value is set by the hypervisor.
+/// The VLEK is generated externally and has to be loaded into the machine.
 ///
 /// The firmware guarantees that the ReportedTcb value is never greater than the installed TCB
 /// version
+///
+/// Since the release of the 1.56 ABI, the Attestation Report was bumped from version 2 to 3.
+/// Due to content differences, both versions are kept separately in order to provide backwards compatibility and most reliable security.
+pub enum AttestationReport {
+    /// Version 2 of the Attestation Report
+    V2(AttestationReportV2),
+    /// Version 3 of the Attestation Report
+    V3(AttestationReportV3),
+}
+
+impl TryFrom<&[u8]> for AttestationReport {
+    type Error = AttestationError;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        // Interpret the version from the first 4 bytes
+        let version = u32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]);
+
+        // Return the appropriate report version
+        match version {
+            2 => {
+                let report: AttestationReportV2 =
+                    bincode::deserialize(bytes).map_err(|e| AttestationError::BincodeError(*e))?;
+                Ok(AttestationReport::V2(report))
+            }
+            3 => {
+                let report: AttestationReportV3 =
+                    bincode::deserialize(bytes).map_err(|e| AttestationError::BincodeError(*e))?;
+                Ok(AttestationReport::V3(report))
+            }
+            _ => Err(AttestationError::UnsupportedReportVersion(version)),
+        }
+    }
+}
+
+#[cfg(any(feature = "openssl", feature = "crypto_nossl"))]
+impl Attestable for AttestationReport {
+    fn signature(&self) -> &Signature {
+        match self {
+            AttestationReport::V2(report) => report.signature(),
+            AttestationReport::V3(report) => report.signature(),
+        }
+    }
+
+    fn measurable_bytes(&self) -> Result<Vec<u8>, io::Error> {
+        match self {
+            AttestationReport::V2(report) => report.measurable_bytes(),
+            AttestationReport::V3(report) => report.measurable_bytes(),
+        }
+    }
+}
+
+/// Trait shared between attestation reports to be able to verify them against the VEK.
+#[cfg(any(feature = "openssl", feature = "crypto_nossl"))]
+trait Attestable {
```

Comment:
Can you define this trait before you `impl` it for `AttestationReport`? Just for readability.